src/server/scene/basic_surface.cpp

```diff
@@ -988,27 +1004,40 @@ auto mir::scene::BasicSurface::content_top_left(State const& state) const -> geo
 
 void mir::scene::BasicSurface::track_outputs()
 {
+    auto on_enter{[&](mg::DisplayConfigurationOutputId const& id) -> bool
+        {
+            if (!tracked_outputs.contains(id))
+            {
+                tracked_outputs.insert(id);
+                return true;
+            }
+            return false;
+        }};
+    auto on_leave{[&](mg::DisplayConfigurationOutputId const& id) -> bool
+        {
+            if (tracked_outputs.contains(id))
+            {
+                tracked_outputs.erase(id);
+                return true;
+            }
+            return false;
+        }};
+
     auto const state{synchronised_state.lock()};
-    std::set<mg::DisplayConfigurationOutputId> tracked;
 
     display_config->for_each_output(
         [&](mg::DisplayConfigurationOutput const& output)
         {
-            if (output.valid() && output.extents().overlaps(state->surface_rect))
+            if (output.valid() && output.used && output.extents().overlaps(state->surface_rect))
             {
                 if (!tracked_outputs.contains(output.id))
                 {
-                    observers->entered_output(this, output.id);
+                    observers->entered_output(this, output.id, on_enter);
```

Comment:
Here we are passing a lambda to a observer that may run on another thread. But that lambda does nothing to synchronise access `tracked_outputs`.

It also seems odd that we are testing `!tracked_outputs.contains(output.id)` before the call and then, in the lambda repeating the same test before updating `tracked_outputs`.

I feel there must be a cleaner solution to whatever problem this is intended to solve. (What does it solve?)

Reply:
I think they are all run in the wayland thread, but will check.

The double checks are needed because we may have multiple `entered_output` and `left_output` notifications. For instance, it may happen that, between the time an `entered_output` notification is emitted and handled, other `entered_output` notifications were handled.

Reply:
Well, actually, only the checks in the hooks are needed for properly functioning. The checks in the `tracked_outputs` could be removed. However, by doing so, most notifications would be spurious.

Reply:
Hmm, I think I'll remove these checks in `tracked_outputs`. They will simplify syncing access to `tracked_outputs` across threads.

Reply:
Okay, done: removed the double checks, and added a mutex for tracked_outputs.

Reply:
This is all moot - we shouldn't be storing the state of frontend state in scene